LML.NPOManagement.Bll/Services/UserService.cs

```diff
@@ -263,5 +263,44 @@ public async Task<bool> ModifyUserInfo(UserInformationModel userInformationModel
             await _dbContext.SaveChangesAsync();
             return true;
         }
+
+        public async Task<List<UserInformationModel>> GetUserByUsername(string userName, bool showGroupsOnly, int userId)
+        {
+            List<User> users = new();
+
+            if (!showGroupsOnly)
+            {
+                users = await _dbContext.Users.ToListAsync();
+            }
+            else
+            {
+                var groupIds = _dbContext.Users
+                    .Where(u => u.Id == userId)
+                    .SelectMany(u => u.Groups.Select(g => g.Id))
+                    .ToList();
+
+                users = _dbContext.Users
+                    .Where(u => u.UsersGroups.Any(g => groupIds.Contains(g.Id)))
+                    .ToList();
+            }
+                
+            List<UserInformationModel> result = new();
+
+            foreach (var user in users)
+            {
+                var userInfo = await _dbContext.UserInformations.Where(u => u.UserId == user.Id).FirstOrDefaultAsync();
+                if (userInfo != null)
+                {
+                    string fullName = userInfo.FirstName + " " + userInfo.LastName;
+                    if (fullName.StartsWith(userName, StringComparison.OrdinalIgnoreCase))
+                    {
+                        var userInformationModel = _mapper.Map<UserInformation, UserInformationModel>(userInfo);
+                        result.Add(userInformationModel);
+                    }
```

Comment:
Research how EF works (debug to verify). When you convert all users to list, it should technically iterate through your entire users table and generate "select *" type of SQL. When you filter the users you want to retrieve by providing lambda, EF should be smart enough to generate SSL query to filter only those records in the database, so no huge data transfer happens from DB to API. When you can't control the number of records in some table (users is a good example), always be extremely careful;